File: fc-chat/demo/part3/ch4/message-system-add-push/src/main/java/net/prostars/messagesystem/handler/websocket/JoinRequestHandler.java
package net.prostars.messagesystem.handler.websocket;

import java.util.Optional;
import net.prostars.messagesystem.constant.IdKey;
import net.prostars.messagesystem.constant.MessageType;
import net.prostars.messagesystem.constant.ResultType;
import net.prostars.messagesystem.dto.domain.Channel;
import net.prostars.messagesystem.dto.domain.UserId;
import net.prostars.messagesystem.dto.websocket.inbound.JoinRequest;
import net.prostars.messagesystem.dto.websocket.outbound.ErrorResponse;
import net.prostars.messagesystem.dto.websocket.outbound.JoinResponse;
import net.prostars.messagesystem.service.ChannelService;
import net.prostars.messagesystem.service.ClientNotificationService;
import org.springframework.data.util.Pair;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketSession;

@Component
@SuppressWarnings("unused")
public class JoinRequestHandler implements BaseRequestHandler<JoinRequest> {

  private final ChannelService channelService;
  private final ClientNotificationService clientNotificationService;

  public JoinRequestHandler(
      ChannelService channelService, ClientNotificationService clientNotificationService) {
    this.channelService = channelService;
    this.clientNotificationService = clientNotificationService;
  }

  @Override
  public void handleRequest(WebSocketSession senderSession, JoinRequest request) {
    UserId senderUserId = (UserId) senderSession.getAttributes().get(IdKey.USER_ID.getValue());

    Pair<Optional<Channel>, ResultType> result;
    try {
      result = channelService.join(request.getInviteCode(), senderUserId);
    } catch (Exception ex) {
      clientNotificationService.sendMessage(
          senderSession,
          senderUserId,
          new ErrorResponse(MessageType.JOIN_REQUEST, ResultType.FAILED.getMessage()));
      return;
    }

    result
        .getFirst()
        .ifPresentOrElse(
            channel ->
                clientNotificationService.sendMessage(
                    senderSession,
                    senderUserId,
                    new JoinResponse(channel.channelId(), channel.title())),
            () ->
                clientNotificationService.sendMessage(
                    senderSession,
                    senderUserId,
                    new ErrorResponse(MessageType.JOIN_REQUEST, result.getSecond().getMessage())));
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/handler/websocket/JoinRequestHandler.kt
package com.yscorp.withpush.messagesystem.handler.websocket

import com.yscorp.withpush.messagesystem.constant.IdKey
import com.yscorp.withpush.messagesystem.constant.MessageType
import com.yscorp.withpush.messagesystem.constant.ResultType
import com.yscorp.withpush.messagesystem.dto.domain.UserId
import com.yscorp.withpush.messagesystem.dto.websocket.inbound.JoinRequest
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.ErrorResponse
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.JoinResponse
import com.yscorp.withpush.messagesystem.service.ChannelService
import com.yscorp.withpush.messagesystem.service.ClientNotificationService
import org.springframework.stereotype.Component
import org.springframework.web.socket.WebSocketSession

@Component
@Suppress("unused")
class JoinRequestHandler(
    private val channelService: ChannelService,
    private val clientNotificationService: ClientNotificationService
) : BaseRequestHandler<JoinRequest> {

    /**
     * 초대코드를 기반으로 채널에 참여 요청을 처리합니다.
     * 성공 시 JoinResponse, 실패 시 ErrorResponse 를 클라이언트에게 전송합니다.
     */
    override fun handleRequest(senderSession: WebSocketSession, request: JoinRequest) {
        val senderUserId = senderSession.attributes[IdKey.USER_ID.value] as? UserId ?: return

        val (channel, result) = try {
            channelService.join(request.inviteCode, senderUserId)
        } catch (ex: Exception) {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                ErrorResponse(MessageType.JOIN_REQUEST, ResultType.FAILED.message)
            )
            return
        }

        if (channel != null) {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                JoinResponse(channel.channelId, channel.title)
            )
        } else {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                ErrorResponse(MessageType.JOIN_REQUEST, result.message)
            )
        }
    }
}
